Bug: Default parameter type mismatch not caught at compile time

## Description

When a function parameter has a default value that doesn't match the declared parameter type, the type checker does not catch this at compile time. The error only surfaces at runtime when the function is called without that argument.

## Reproduction

```ez
import @std
using std

// Default value "hello" doesn't match int type - should FAIL at compile time
do bad_default(x int = "hello") -> int {
    return x
}

do main() {
    // This compiles fine!
    println("Does this compile?")

    // Error only happens here at runtime:
    // println("Result: ${bad_default()}")
}
```

## Expected Behavior

The type checker should emit an error at the function definition:
```
error[E3001]: default parameter type mismatch: parameter 'x' expects int, got string
```

## Actual Behavior

The function definition compiles without error. The type mismatch is only caught at runtime when the function is called and the default value is used.

## Impact

High - invalid function definitions are accepted, leading to runtime errors instead of compile-time errors.

--- a/integration-tests/fail/errors/E3001_default_param_type_mismatch.ez
+++ b/integration-tests/fail/errors/E3001_default_param_type_mismatch.ez
@@ -1,0 +1,12 @@
+/*
+ * Error Test: E3001 - default-param-type-mismatch
+ * Expected: "type mismatch" or "default"
+ */
+
+do bad_default(x int = "hello") -> int {
+    return x
+}
+
+do main() {
+    println("should not compile")
+}

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -799,6 +799,22 @@ func (tc *TypeChecker) checkFunctionDeclaration(node *ast.FunctionDeclaration) {
 				param.Name.Token.Column,
 			)
 		}
+
+		// Check default value type matches parameter type (#582)
+		if param.DefaultValue != nil {
+			defaultType, ok := tc.inferExpressionType(param.DefaultValue)
+			if ok && !tc.typesCompatible(param.TypeName, defaultType) {
+				line, col := tc.getExpressionPosition(param.DefaultValue)
+				tc.addError(
+					errors.E3001,
+					fmt.Sprintf("default value type mismatch: parameter '%s' expects %s, got %s",
+						param.Name.Value, param.TypeName, defaultType),
+					line,
+					col,
+				)
+			}
+		}
+
 		sig.Parameters = append(sig.Parameters, &Parameter{
 			Name:       param.Name.Value,
 			Type:       param.TypeName,

